### ISSUE TITLE
invoiceninja + traefik = ERR_TOO_MANY_REDIRECTS

### ISSUE THREAD

**Issue by w4tzmann** [invoiceninja/dockerfiles#180]
Hi,

is there someone who can help me running invoiceninja behind traefik?

Other apps (e .g. mailcow, grafana, portainer) are working fine. But when i try to open invoiceninja in the browser i only get "ERR_TOO_MANY_REDIRECTS"

Thats my docker-compose.yml:

```
version: '3.7'

services:
  server:
    image: caddy:alpine
    restart: always
    environment:
      - APP_URL=https://invoiceninja.tld
    volumes:
      # Vhost configuration
      - ./config/caddy/Caddyfile:/etc/caddy/Caddyfile

      # Configure your mounted directories, make sure the folder 'public' and 'storage'
      # exist, before mounting them
      #-  public:/var/www/app/public
      #-  storage:/var/www/app/storage
      # you may use a bind-mounted host directory instead, so that it is harder to accidentally remove the volume and lose all your data!
      - ./data/public:/var/www/app/public:rw,delegated
      - ./data/storage:/var/www/app/storage:rw,delegated
    depends_on:
      - app
    # Run webserver nginx on port 80
    # Feel free to modify depending what port is already occupied
    #ports:
     # - "80:80"
     # - "443:443"
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.invoiceninja.entrypoints=http"
      - "traefik.http.routers.invoiceninja.rule=Host(`invoiceninja.tld`)"  ## Hier eure Domain eingeben ##
      - "traefik.http.middlewares.invoiceninja-https-redirect.redirectscheme.scheme=https"
      - "traefik.http.routers.invoiceninja.middlewares=invoiceninja-https-redirect"
      - "traefik.http.routers.invoiceninja-secure.entrypoints=https"
      - "traefik.http.routers.invoiceninja-secure.rule=Host(`invoiceninja.tld`)" ## Hier eure Domain eingeben ##
      - "traefik.http.routers.invoiceninja-secure.tls=true"
      - "traefik.http.routers.invoiceninja-secure.service=invoiceninja"
      - "traefik.http.services.invoiceninja.loadbalancer.server.port=80"
      - "traefik.http.routers.invoiceninja.middlewares=secHeaders@file"
      - "traefik.docker.network=proxy"
    networks:
      - invoiceninja
      - proxy

  app:
    image: invoiceninja/invoiceninja:5
    restart: always
    environment:
      - APP_URL=https://invoiceninja.tld
      - APP_KEY=%key%
      - MULTI_DB_ENABLED=false
      - DB_HOST1=db
    volumes:
      # Configure your mounted directories, make sure the folder 'public' and 'storage'
      # exist, before mounting them
      #-  public:/var/www/app/public
      #-  storage:/var/www/app/storage
      # you may use a bind-mounted host directory instead, so that it is harder to accidentally remove the volume and lose all your data!
      - ./data/public:/var/www/app/public:rw,delegated
      - ./data/storage:/var/www/app/storage:rw,delegated
    depends_on:
      - db
    networks:
      - invoiceninja

  db:
    image: mysql:5
    restart: always
    environment:
      - MYSQL_ROOT_PASSWORD=ninjaAdm1nPassword
      - MYSQL_USER=ninja
      - MYSQL_PASSWORD=ninja
      - MYSQL_DATABASE=db-ninja-01
    volumes:
      #- mysql-data:/var/lib/mysql:rw
      # you may use a bind-mounted host directory instead, so that it is harder to accidentally remove the volume and lose all your data!
      - ./data/mysql/data:/var/lib/mysql:rw,delegated
    networks:
      - invoiceninja


  # cron:
  #   image: invoiceninja/invoiceninja
  #   volumes:
  #     -  storage:/var/www/app/storage
  #     -  logo:/var/www/app/public/logo
  #     -  public:/var/www/app/public
  #   entrypoint: |
  #     bash -c 'bash -s <<EOF
  #     trap "break;exit" SIGHUP SIGINT SIGTERM
  #     sleep 300s
  #     while /bin/true; do
  #       ./artisan ninja:send-invoices
  #       ./artisan ninja:send-reminders
  #       sleep 1d
  #     done
  #     EOF'
  #   networks:
  #     - invoiceninja

volumes:
  mysql-data:
  public:
  storage:

networks:
  invoiceninja:
  proxy:
   external: true
```

Thank you in advance!

**Comment by codedge** [comment]
Sorry, I have no experience with trafik. I guess this is a problem with redirects between HTTP -> HTTPS. Unfortunately I cannot provide any solution for this. Maybe someone else has more experience.

**Comment by w4tzmann** [comment]
Hey,

thanks for your reply. I was playing around with the Caddyfile:
```
{
auto_https disable_redirects
}
{$APP_URL} {
    root * /var/www/app/public
    php_fastcgi app:9000
    encode zstd gzip
    file_server browse
}
```
And APP_URL with https give me a Bad Gateway. I have no public dns entry fore the domain, so acme challange failed.
Log:

> {"level":"info","ts":1596885077.838202,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"},
> {"level":"info","ts":1596885077.8403993,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]},
> 2020/08/08 11:11:17 [INFO][cache:0xc00003bce0] Started certificate maintenance routine,
> {"level":"info","ts":1596885077.8406887,"logger":"http","msg":"server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS","server_name":"srv0","https_port":443},
> {"level":"info","ts":1596885077.8420527,"logger":"tls","msg":"cleaned up storage units"},
> {"level":"info","ts":1596885077.8423533,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["invoiceninja.rz-walhall.de"]},
> {"level":"info","ts":1596885077.8435454,"msg":"autosaved config","file":"/config/caddy/autosave.json"},
> {"level":"info","ts":1596885077.8436892,"msg":"serving initial configuration"},
> 2020/08/08 11:11:17 [INFO][invoiceninja.rz-walhall.de] Obtain certificate; acquiring lock...,
> 2020/08/08 11:11:17 [INFO][invoiceninja.rz-walhall.de] Obtain: Lock acquired; proceeding...,
> 2020/08/08 11:11:18 [INFO] [invoiceninja.rz-walhall.de] acme: Obtaining bundled SAN certificate given a CSR,
> 2020/08/08 11:11:18 [INFO][invoiceninja.rz-walhall.de] Waiting on rate limiter...,
> 2020/08/08 11:11:18 [INFO][invoiceninja.rz-walhall.de] Done waiting,
> 2020/08/08 11:11:19 [INFO] [invoiceninja.rz-walhall.de] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/6394492763,
> 2020/08/08 11:11:19 [INFO] [invoiceninja.rz-walhall.de] acme: Could not find solver for: tls-alpn-01,
> 2020/08/08 11:11:19 [INFO] [invoiceninja.rz-walhall.de] acme: use http-01 solver,
> 2020/08/08 11:11:19 [INFO] [invoiceninja.rz-walhall.de] acme: Trying to solve HTTP-01,
> 2020/08/08 11:11:25 [INFO] Deactivating auth: https://acme-v02.api.letsencrypt.org/acme/authz-v3/6394492763,
> 2020/08/08 11:11:25 [INFO] Unable to deactivate the authorization: https://acme-v02.api.letsencrypt.org/acme/authz-v3/6394492763,
> 2020/08/08 11:11:25 [ERROR] error: one or more domains had a problem:,
> [invoiceninja.rz-walhall.de] acme: error: 400 :: urn:ietf:params:acme:error:dns :: DNS problem: NXDOMAIN looking up A for invoiceninja.rz-walhall.de - check that a DNS record exists for this domain, url: ,
>  (challenge=http-01 remaining=[tls-alpn-01]),
> 2020/08/08 11:11:27 [INFO] [invoiceninja.rz-walhall.de] acme: Obtaining bundled SAN certificate given a CSR,
> 2020/08/08 11:11:28 [INFO] [invoiceninja.rz-walhall.de] AuthURL: https://acme-v02.api.letsencrypt.org/acme/authz-v3/6394494522,
> 2020/08/08 11:11:28 [INFO] [invoiceninja.rz-walhall.de] acme: use tls-alpn-01 solver,
> 2020/08/08 11:11:28 [INFO] [invoiceninja.rz-walhall.de] acme: Trying to solve TLS-ALPN-01,
> 2020/08/08 11:11:28 http: TLS handshake error from 127.0.0.1:36328: EOF,
> 2020/08/08 11:11:34 [INFO] Deactivating auth: https://acme-v02.api.letsencrypt.org/acme/authz-v3/6394494522,
> 2020/08/08 11:11:34 [ERROR] error: one or more domains had a problem:,
> [invoiceninja.rz-walhall.de] acme: error: 400 :: urn:ietf:params:acme:error:dns :: DNS problem: NXDOMAIN looking up A for invoiceninja.rz-walhall.de - check that a DNS record exists for this domain, url: ,
>  (challenge=tls-alpn-01 remaining=[]),
> 2020/08/08 11:11:34 [INFO] Unable to deactivate the authorization: https://acme-v02.api.letsencrypt.org/acme/authz-v3/6394494522,
> 2020/08/08 11:11:36 [ERROR] attempt 1: [invoiceninja.rz-walhall.de] Obtain: [invoiceninja.rz-walhall.de] error: one or more domains had a problem:,
> [invoiceninja.rz-walhall.de] acme: error: 400 :: urn:ietf:params:acme:error:dns :: DNS problem: NXDOMAIN looking up A for invoiceninja.rz-walhall.de - check that a DNS record exists for this domain, url: ,
>  - retrying in 1m0s (19.063486126s/720h0m0s elapsed)...,

```
{
auto_https off
}
{$APP_URL} {
    root * /var/www/app/public
    php_fastcgi app:9000
    encode zstd gzip
    file_server browse
}
```

Also result in a Bad Gateway message. Docker-Log for server says:

> {"level":"info","ts":1596884893.8438277,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"},
> {"level":"info","ts":1596884893.8537297,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["127.0.0.1:2019","localhost:2019","[::1]:2019"]},
> 2020/08/08 11:08:13 [INFO][cache:0xc000698900] Started certificate maintenance routine,
> {"level":"info","ts":1596884893.8566506,"logger":"tls","msg":"cleaned up storage units"},
> {"level":"info","ts":1596884893.8569112,"msg":"autosaved config","file":"/config/caddy/autosave.json"},
> {"level":"info","ts":1596884893.8569236,"msg":"serving initial configuration"},

If i try auto_https disable_redirects and APP_URL http i get 500 | Server Error. Log says:

> {"level":"info","ts":1596885255.599783,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"},
> {"level":"info","ts":1596885255.6018233,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]},
> {"level":"info","ts":1596885255.6020827,"logger":"http","msg":"server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server","server_name":"srv0","http_port":80},
> {"level":"info","ts":1596885255.6034842,"logger":"tls","msg":"cleaned up storage units"},
> {"level":"info","ts":1596885255.6038184,"msg":"autosaved config","file":"/config/caddy/autosave.json"},
> {"level":"info","ts":1596885255.6038563,"msg":"serving initial configuration"},
> 2020/08/08 11:14:15 [INFO][cache:0xc00035ed80] Started certificate maintenance routine,

And auto_https off and APP_URL brings also 500 | Server Error
Log:

> {"level":"info","ts":1596885414.8367686,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"},
> {"level":"info","ts":1596885414.8389795,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]},
> {"level":"info","ts":1596885414.8403049,"logger":"tls","msg":"cleaned up storage units"},
> 2020/08/08 11:16:54 [INFO][cache:0xc00003b3e0] Started certificate maintenance routine,
> {"level":"info","ts":1596885414.841105,"msg":"autosaved config","file":"/config/caddy/autosave.json"},
> {"level":"info","ts":1596885414.8411658,"msg":"serving initial configuration"},

Any ideas for me?

**Comment by digitalvyce** [comment]
I had the same issue running IN behind traefik, I was able to get it to work by foring caddy to not do letsencry/tls (since Traefik is handling that for us). Because of this, it is also important to configure Invoiceninja to not use https when you get to the setup page. But the URL should still have the HTTPS (at least thats what I understand, if someone else can confirm)

I was able to tell caddy to not not do https by declaring http in the url in the Caddyfile 
example 

`http://sub.domain.com {
    root * /var/www/app/public
    php_fastcgi invoiceninja:9000
    encode zstd gzip
    file_server browse
}`

(I renamed my containers in my docker compose so my invoiceninja:9000 would be y our app:9000)

You can confirm it is turned of by reading the log file of the Caddy container, and you'll see it mention disabling auto-https

If you are using Cloudflare, be sure not to proxy it (grey cloud) at least until you have it working with out it, then turn it on. If the issue reproduces with the Cloudflare proxy turned on, you may need to switch the SSL settings of the domain from Full (Strict) to just Full

I was able to get InvoiceNinja to work, but had an issue with email invoices, and another issue the cron jobs being configured. Im unsure if Invoice ninja 5 is using the cron container as shown in their docker-compose example, or if its commented out because invoiceninja 5 container is now doing it within it? Either way, hope this helps

**Comment by w4tzmann** [comment]
Hi,

thanks for your answer.
I think the http thing is solved. My problem is the php 500 error now...

**Comment by codedge** [comment]
Just to let you know: There is also a [guide](https://github.com/invoiceninja/dockerfiles/wiki) now which uses nginx as webserver without TLS.

**Comment by codedge** [comment]
@digitalvyce I created a [wiki page](https://github.com/invoiceninja/dockerfiles/wiki/Cronjobs) to describe the current situation with the cronjobs.

**Comment by w4tzmann** [comment]
Hi @codedge,

Thanks for the guide!
I followed the instructions. But i ran into 2 problems:
Mounted volumes give me this error: 

> cp: can't create directory '/var/www/app/storage/app': Permission denied

Using `chown -R 1000:82 app` solves this problem.

However, the attempt to access the website fails. The Error Log contains the following line:

> 192.168.96.4 -  18/Aug/2020:15:12:23 +0000 "GET /index.php" 500

So the original problem with traffic and ERR_TO_MANY_REDIRECTS seems to be solved. Should I open a new issue?

**Comment by codedge** [comment]
Hmm.. that's odd.

Can you please add `APP_DEBUG=true` to you environment vars of the `app` (invoiceninja) container? That should give you a complete output of the error in the browser.

> Should I open a new issue?
That would be superb!

**Comment by w4tzmann** [comment]
So then let's go on: #190
